# Progetto del corso - Introduzione al Pensiero Computazionale e alla Data Science

**Dataset scelto:** Employee Attrition
**Obiettivo:** previsione dell'abbandono dei dipendenti (`Attrition`)

**Team:** [Paolo Magnanelli - 1231652], [Asia Milan - matricola].

---
# Fase 1 - Setup del progetto

In questa cella importiamo le librerie che useremo in tutto il notebook.

- `pandas`, `numpy` per la gestione dei dati
- `matplotlib`, `seaborn` per le visualizzazioni
- `scikit-learn` per modellazione e metriche

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    ConfusionMatrixDisplay,
    classification_report,
    precision_score,
    recall_score,
    f1_score,
)

In [8]:
df = pd.read_csv("Employee_Attrition.csv")
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,Y,No,23,4,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,Y,Yes,15,3,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2


---
# Fase 2 - Descrizione e comprensione del dataset

Lo schema che useremo per ogni domanda è:

1. esposizione della domanda/ipotesi;
2. recupero e visualizzazione dell'informazione;
3. risposta/considerazione.

**Iniziamo con una panoramica**: dimensioni, tipi di variabili e presenza di
valori mancanti.

In [9]:
print("Numero di righe:", df.shape[0])
print("Numero di colonne:", df.shape[1])
df.info()

Numero di righe: 1470
Numero di colonne: 35
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non

**Considerazioni:** il dataset contiene 1470 osservazioni (dipendenti) e 35
colonne. Non sono presenti tipi di dato anomali a prima vista: la maggior parte
delle variabili è numerica (int64), mentre alcune sono categoriche (stringhe).
Procediamo a verificare i valori mancanti.

### Domanda 1 - Bla Bla?

### Domanda 2 - Bla Bla?


### Domanda 3 - Bla Bla?


### Domanda 4 - Bla Bla?


### Domanda 5 - Bla Bla?


### Domanda 6 - Bla Bla?


### Domanda 7 - Bla Bla?


### Domanda 8 - Bla Bla?


### Conclusioni della fase 2

---
# Fase 3 - Analisi esplorativa e visualizzazione

**Obiettivo:** capire (e comunicare) le relazioni tra le feature, in particolare
rispetto al target `Attrition`.

Per facilitare i confronti, creiamo una versione numerica del target
(`Attrition_num`): `1` = "Yes" (il dipendente ha lasciato l'azienda), `0` = "No".

In [10]:
df["Attrition_num"] = (df["Attrition"] == "Yes").astype(int)
df[["Attrition", "Attrition_num"]].head()

,Attrition,Attrition_num
0,Yes,1
1,No,0
2,Yes,1
3,No,0
4,No,0


### Domanda 1 - Bla Bla?

### Domanda 2 - Bla Bla?

### Domanda 3 - Bla Bla?

### Domanda 4 - Bla Bla?

---
# Fase 4 - Modellazione

---
# Fase 5 - Valutazione e interpretazione dei risultati

---
# Conclusioni e possibili approfondimenti